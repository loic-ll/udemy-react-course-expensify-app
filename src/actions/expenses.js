import {database} from '../firebase/firebase';

const expensesRef = database.ref('expenses');

const defaultValues = {
  amount: 0,
  createdAt: 0,
  description: '',
  note: '',
};

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expense = defaultValues) => {
  return dispatch => {
    return expensesRef.push(expense).then(ref => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        }),
      );
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, values) => ({
  type: 'EDIT_EXPENSE',
  id,
  values,
});

export const startEditExpense = (id, values) => {
  return dispatch => {
    return expensesRef
      .child(id)
      .update(values)
      .then(() => {
        dispatch(editExpense(id, values));
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = id => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const startRemoveExpense = id => {
  return dispatch => {
    return expensesRef
      .child(id)
      .remove()
      .then(() => {
        dispatch(removeExpense(id));
      });
  };
};

// SET_EXPENSES

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => {
  return dispatch => {
    return expensesRef.once('value').then(snapshot => {
      const expenses = [];
      snapshot.forEach(childsnap => {
        expenses.push({
          id: childsnap.key,
          ...childsnap.val(),
        });
      });

      dispatch(setExpenses(expenses));
    });
  };
};
