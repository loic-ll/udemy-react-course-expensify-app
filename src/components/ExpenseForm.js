import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    let state = (this.state = {
      description: '',
      note: '',
      amount: '',
      error: '',
      createdAt: moment(),
      calendarFocused: false,
    });

    if (props.expense) {
      const expense = props.expense;
      state.description = expense.description;
      state.note = expense.note;
      state.amount = (expense.amount / 100).toString();
      state.createdAt = moment(expense.createdAt);
    }
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({note}));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount}));
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({createdAt}));
    }
  };
  onFocusChange = ({focused: calendarFocused}) => {
    this.setState(() => ({calendarFocused}));
  };
  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'Please provide all informations'}));
    } else {
      this.setState(() => ({error: undefined}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          autoFocus
          className="text-input"
          name="description"
          onChange={this.onDescriptionChange}
          placeholder="Description"
          type="text"
          value={this.state.description}
        />
        <input
          className="text-input"
          name="amount"
          onChange={this.onAmountChange}
          placeholder="Amount"
          step="0.01"
          type="number"
          value={this.state.amount}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          focused={this.state.calendarFocused}
          isOutsideRange={() => false}
          numberOfMonths={1}
          onDateChange={this.onDateChange}
          onFocusChange={this.onFocusChange}
        />
        <textarea
          className="textarea"
          name="note"
          onChange={this.onNoteChange}
          placeholder="Add a note (optional)"
          value={this.state.note}
        />
        <div>
          <button className="button">Save</button>
        </div>
      </form>
    );
  }
}
