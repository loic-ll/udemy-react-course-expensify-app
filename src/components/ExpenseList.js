import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-on-mobile">Expenses</div>
      <div className="show-on-desktop">Expense</div>
      <div className="show-on-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No expenses</span>
        </div>
      ) : (
        props.expenses.map((expense, index) => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )}
    </div>
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
export {ExpenseList};
