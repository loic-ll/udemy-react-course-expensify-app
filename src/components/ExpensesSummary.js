import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import getTotal from '../selectors/expenses-total'

const ExpensesSummary = ({count, total}) => {
  const formattedTotal = numeral(total / 100).format('0,0.00 $');
  return (
    <div>
      {
        (count === 0) ? (<h2>No expenses</h2>) : (
          (count === 1) ? (<h2>1 expense: {formattedTotal}</h2>) :
          (<h2>{count} expenses totalling: {formattedTotal}</h2>) )
      }
    </div>
  );
};

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses, state.filters);

  return {
    count: expenses.length,
    total: getTotal(expenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
export { ExpensesSummary }
