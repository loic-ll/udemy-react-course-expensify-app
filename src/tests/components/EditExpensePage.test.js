import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let editExpense, history, removeExpense, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage 
      expense={expenses[0]}
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history} />);
});

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(editExpense).toHaveBeenCalledWith(expenses[0].id, expenses[0]);
  expect(history.push).toHaveBeenCalledWith('/');
});

test('should handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenCalledWith({ id: expenses[0].id });
  expect(history.push).toHaveBeenCalledWith('/');
});
