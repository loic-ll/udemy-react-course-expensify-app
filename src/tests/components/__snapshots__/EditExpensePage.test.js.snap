// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 0,
        "createdAt": 0,
        "description": "abc",
        "id": "0",
        "note": "now",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
